### PR TITLE
Fix #35 and improve testing setup

### LINKED ISSUE #35: Command parser broken
I'm afraid the issue I raised in https://github.com/enen92/script.tubecast/pull/32#issuecomment-581199670 regarding the string replacements in the command parser already caught up with us.
I started working on some unit tests in [siku2/tests](https://github.com/siku2/script.tubecast/tree/tests) and the following payload breaks the command parser:

```python
{"devices": "[{\"app\":\"kodi-tubecast\",\"capabilities\":\"que,mus\",\"clientName\":\"unknown\",\"experiments\":\"\",\"name\":\"Kodi (kodi)\",\"id\":\"c8277ac4-ke86-4f8b-8fe2-1236bef43397\",\"type\":\"LOUNGE_SCREEN\",\"hasCc\":\"true\"}]","connectionEventDetails":"{\"deviceId\":\"c8277ac4-ke86-4f8b-8fe2-1236bef43397\"}"}
```

After the string substitution it turns into:

```python
{"devices": "[{"app":"kodi-tubecast","capabilities":"que,mus","clientName":"unknown","experiments":"","name":"Kodi (kodi)","id":"c8277ac4-ke86-4f8b-8fe2-1236bef43397","type":"LOUNGE_SCREEN","hasCc":"true"}]","connectionEventDetails":{"deviceId":"c8277ac4-ke86-4f8b-8fe2-1236bef43397"}}
```

Notice how `"[` remains untouched which causes severe quoting issues.

This isn't a payload deliberately designed to break the parser, it's the `loungeStatus` command which is one of the initial commands.
Before the changes to the command parser this command was ignored entirely which is why it hasn't been an issue so far.

### PR DESCRIPTION
**Pull request by siku2**
Fixes #35.

- Added mock modules for `xbmc`, `xbmcaddon`, and `xbmcgui`. This makes it possible to test code that imports Kodi modules. This will allow us to add unit tests for other parts of the code base in the future.
- Switched back to the command version prior to your fixes (apart from the Python 3 compatibility)<sup class="footnote-ref"><a href="#fn1" id="fnref1">[1]</a></sup>.
- Unit tests for more commands

<section class="footnotes">
<ol class="footnotes-list">
<li id="fn1"  class="footnote-item"><p>This includes suppressing exceptions. I thought about it and considering that #35 would have been half as bad had it not crashed the listener I think it's the right call.
In addition to logging the exception we could also show a notification to the user. This would make it easier to catch broken commands.<a href="#fnref1" class="footnote-backref">↩</a></p>
</li>
</ol>
</section>

### CODE REVIEW

File: .github/workflows/py-test.yml [hunk 1]
```diff
@@ -20,5 +20,5 @@ jobs:
     - name: Test with pytest
       run: |
         python -m pip install --upgrade pip
-        pip install pytest
+        pip install -r test-requirements.txt
```
Comment (enen92) [inline .github/workflows/py-test.yml]: please add as export-ignore to .gitattributes. We use that to create distribution zips and submit the PR to kodi repo

File: tests/test_cmd_parser.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+cmd_remote_connected = '550\n[[6,["remoteConnected",{"app":"android-phone","pairingType":"dial","capabilities":"que,atp,mus","ui":"true","clientName":"android","experiments":"","name":"name","remoteControllerUrl":"","id":"id","type":"REMOTE_CONTROL","device":"{\\"app\\":\\"android-phone\\",\\"pairingType\\":\\"dial\\",\\"capabilities\\":\\"que,atp,mus\\",\\"clientName\\":\\"android\\",\\"experiments\\":\\"\\",\\"name\\":\\"name\\",\\"remoteControllerUrl\\":\\"\\",\\"id\\":\\"id\\",\\"type\\":\\"REMOTE_CONTROL\\"}"}]]\n]\n'
+cmd_remote_disconnected = '554\n[[79,["remoteDisconnected",{"app":"android-phone-14.21.54","pairingType":"dial","capabilities":"que,atp,mus","ui":"true","clientName":"android","experiments":"","name":"name","remoteControllerUrl":"","id":"id","type":"REMOTE_CONTROL","device":"{\\"app\\":\\"android-phone\\",\\"pairingType\\":\\"dial\\",\\"capabilities\\":\\"que,atp,mus\\",\\"clientName\\":\\"android\\",\\"experiments\\":\\"\\",\\"name\\":\\"name\\",\\"remoteControllerUrl\\":\\"\\",\\"id\\":\\"id\\",\\"type\\":\\"REMOTE_CONTROL\\"}"}]]\n]\n'
```
Comment (enen92) [inline tests/test_cmd_parser.py]: please add the same test commands we had on test_cmds.py. I added them for a reason, they are real edge cases I received from the youtube socket

Comment (siku2) [reply tests/test_cmd_parser.py]: This is rather problematic... Let me try to unwrap this:

The only difference between `cmd_1` and `cmd_2` is that the latter doesn't have the opening square brackets.
The commands are from the time the listener read the commands line by line so I think it's likely that the bracket was on the previous line:
```
[\n
20,["remoteDisconnected", ...
```

Then there's the other thing, the difference between the two commands and the one I included in the new tests. The difference is that the old commands use `\"` instead of `\\"`.
I find it really hard to explain this, so please bear with me...

Open a new python terminal and paste `cmd_1` (including the single quotes). You'll find that it turns into this:
```
'[[20,["remoteDisconnected",{"app":"android-phone-13.05.52","pairingType":"dial","capabilities":"atp,que,mus","ui":"true","clientName":"android","experiments":"","name":"MYPHONE","remoteControllerUrl":"ws://","id":"something","type":"REMOTE_CONTROL","device":"{"app":"android-phone-13.05.52","pairingType":"dial","capabilities":"atp,que,mus","clientName":"android","experiments":"","name":"MYPHONE","remoteControllerUrl":"ws://","id":"id","type":"REMOTE_CONTROL"}"}]]'
``` 
Notice how `\"` turned into `"`? That is because `' \" ' == ' " '` ([Python Reference Manual on String literals](https://docs.python.org/2.0/ref/strings.html)).
So why does the command use `\"` in the first place? Well, if you print `\\"` in Python you'll find that's exactly what happens. `\\"` turns into `\"`.
So what I think happened is that you took the command straight from the logs without accounting for the fact that you need to escape `\"` to `\\"`.

I'm sorry if this sounded condescending, I just wanted to make sure to get the point across.

If you're 100% sure that I'm wrong in both cases then we have a bit of a problem on our hands.
Take a look at the code snippet I posted above, there's no way we could parse something like `"{"app":"android-phone-13.05.52", ...}"` properly. I find it hard to believe that the YouTube app could produce such an output under any circumstances.

Comment (enen92) [reply tests/test_cmd_parser.py]: Sorry for the delay, got busy with real life. I am fine with dropping those 2 tests but can you at least leave the old lines there (commented) just for reference? Thanks

**Comment by enen92** [comment]
Thanks much! I'll submit an update to repo.
